src/probability_theory/density.lean

```diff
@@ -0,0 +1,380 @@
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :
+  measurable (pdf X ℙ μ) :=
+begin
+  by_cases hX : has_pdf X ℙ μ,
+  { rw [pdf, dif_pos hX],
+    exact (classical.some_spec hX.pdf').1 },
+  { rw [pdf, dif_neg hX],
+    exact measurable_zero }
+end
+
+lemma pdf_spec {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ] :
+  measure.map X ℙ = μ.with_density (pdf X ℙ μ) :=
+begin
+  rw [pdf, dif_pos hX],
+  exact (classical.some_spec hX.pdf').2
+end
+
+lemma pdf_spec' {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ]
+  {s : set E} (hs : measurable_set s) :
+  measure.map X ℙ s = ∫⁻ x in s, pdf X ℙ μ x ∂μ :=
+by rw [← with_density_apply _ hs, pdf_spec X ℙ μ]
+
+namespace pdf
+
+variables {m : measurable_space α} {ℙ : measure α} {μ : measure E}
+
+section
+
+lemma lintegral_eq_measure_univ {X : α → E} [has_pdf X ℙ μ] (hX : measurable X) :
+  ∫⁻ x, pdf X ℙ μ x ∂μ = ℙ set.univ :=
+begin
+  rw [← set_lintegral_univ, ← pdf_spec' X ℙ μ measurable_set.univ,
+      measure.map_apply hX measurable_set.univ, set.preimage_univ],
+end
+
+lemma ae_lt_top {m : measurable_space α} (ℙ : measure α) [is_finite_measure ℙ] {μ : measure E}
+  {X : α → E} (hX : measurable X) : ∀ᵐ x ∂μ, pdf X ℙ μ x < ∞ :=
+begin
+  by_cases hpdf : has_pdf X ℙ μ,
+  { refine ae_lt_top (measurable_pdf X ℙ μ) _,
+    rw lintegral_eq_measure_univ hX,
+    { exact (measure_lt_top _ _).ne },
+    { exact hpdf } },
+  { rw [pdf, dif_neg hpdf],
+    exact filter.eventually_of_forall (λ x, with_top.zero_lt_top) }
+end
+
+lemma of_real_to_real_ae_eq [is_finite_measure ℙ] {X : α → E} (hX : measurable X) :
+  (λ x, ennreal.of_real (pdf X ℙ μ x).to_real) =ᵐ[μ] pdf X ℙ μ :=
+begin
+  by_cases hpdf : has_pdf X ℙ μ,
+  { haveI := hpdf,
+    change μ {x : E | ennreal.of_real (pdf X ℙ μ x).to_real ≠ pdf X ℙ μ x} = 0,
+    have : ∀ x, ennreal.of_real (pdf X ℙ μ x).to_real ≠ pdf X ℙ μ x ↔
+      pdf X ℙ μ x = ∞,
+    { intro x, split; intro h,
+      { by_contra htop,
+        rw [← ne.def, ← lt_top_iff_ne_top] at htop,
+        exact h (ennreal.of_real_to_real htop.ne) },
+      { rw h, simp } },
+    { simp_rw this,
+      suffices hne : ∀ᵐ x ∂μ, pdf X ℙ μ x ≠ ∞,
+      { simp_rw [ae_iff, not_not] at hne, exact hne },
+      convert ae_lt_top ℙ hX,
+      simp_rw [lt_top_iff_ne_top] } },
+  { convert ae_eq_refl _,
+    ext1 x,
+    rw [pdf, dif_neg hpdf, pi.zero_apply, ennreal.zero_to_real, ennreal.of_real_zero] }
+end
+
+/-- **The Law of the Unconscious Statistician**: Given a random variable `X` and a measurable
+function `f`, we have `f ∘ X` is a random variable and have expectation `∫ x, f x * pdf X ∂μ`
+where `μ` is a measure on the codomain of `X`. -/
+lemma integral_mul_eq_integral' [is_finite_measure ℙ] {X : α → E} [has_pdf X ℙ μ]
+  (hX : measurable X) (f : E → ℝ) (hf : measurable f)
+  (hpdf : integrable (λ x, f x * (pdf X ℙ μ x).to_real) μ) :
```

Comment:
Is this equivalent to `f ∘X` being integrable? (I have no idea). If yes, that hypothesis can be removed since both integrals would be zero.

Reply:
The condition is now removed :)